Hunterised ros catkin 0.7.11

lsolanka-seebyte opened the pull request:
The first installment for ruslo/hunter#1397.

Catkin is a base dependency for each ROS package, but it is not hunter ready. The current patches do the following:
- Do not extract absolute path names from libraries and targets, instead use exported targets
- Properly find_package packages which are listed as dependencies of each catkin package
- Use exported targets when the package is being installed instead of hard coded paths to libraries
  - This requires the dependent packages to define one variable: PROJECT_INSTALLSPACE_LIBRARIES that lists the exported targets. This will be then added to ${catkin_LIBRARIES}, which is a standard way how ROS pulls libraries for linking. This maximizes backwards compatibility with already existing CMakeLists.txt of ROS packages.
- Clean up some other commands which interfere with BUILD_SHARED_LIBS

The final result is that most users will only need to add `hunter_add_package(ros-ros_comm)` (after all packages are added to Hunter) and not touch any other code in their CMakeLists.txt.

Let me know how this looks.

lsolanka committed:
hunterise

sanitise package config file generation

remove devel space generation

remove default ON setting for BUILD_SHARED_LIBS

find_package both DEPENDS and CATKIN_DEPENDS instead of collating libs

include Targets.cmake file only if catkin_package defines some LIBRARIES

fix typo

reinstance PKG_CFG_EXTRAS in devel space

reinstate devel space with a dummy Targets.cmake file

This file is empty, for now inside a single repository, all packages
must point to the correct targets explicitly.

But they can still call find_package(catkin COMPONENTS ...) even
without a proper install of the dependent package.

catkinConfig.cmake does not unpack libraries

fix wrong file GENERATE command

add DEVELSPACE_LIBRARIES to catkin_package

This is to properly export develspace targets/libs so that the naming
does not contain any namespace prefix (when specifying imported targets)

ditch generation of dummy Targets.cmake file in develspace

fix missing PROJECT_ prefix

reverse the logic of develspace and installspace targets

rename PROJECT_EXPORTED_TARGETS to INSTALLSPACE_LIBRARIES

because the name clashes with EXPORTED_TARGETS catkin_package parameter

remove redundant line

attempt to use hunterised GTest

properly find and add hunterised GTest

fix checking if GMock/GTest was found

use PROJECT_INSTALLSPACE_LIBRARIES instead of INSTALLSPACE_LIBRARIES

release hunterised version

re-instate develspace include dirs

catkinConfig needs to pull {component_INCLUDE_DIRS}

Otherwise when generating C++ headers from messages/services does not
get pulled to devel space

remove unnecessary gtest dependency targets

add missing INCLUDE_DIRS for transitive dependencies

This was removed but it's not correct. Because 2nd-or-higher order dependencies
will not propagate their ${project}_INCLUDE_DIRS to the parent.

lsolanka committed:
clean up some commented out code and formatting to minimize the diff

lsolanka committed:
re-instate PKG_CONFIG commands to minimize diff (not used anyway)

ruslo commented on cmake/platform/windows.cmake:
```diff
@@ -3,11 +3,6 @@
 
-# Make sure this is already defined as a cached variable (@sa tools/libraries.cmake)
-if(NOT DEFINED BUILD_SHARED_LIBS)
-  option(BUILD_SHARED_LIBS "Build dynamically-linked binaries" ON)
```
`option` should work fine, no need to remove it

lsolanka-seebyte replied:
Both platform/windows.cmake and tools/libraries.cmake are included with every catkin package CMakeLists.txt. When I leave this option with default `ON` the system tries to build everything with shared libraries. Perhaps I'm missing something? Is `BUILD_SHARED_LIBS` set to `OFF` by default inside `HunterGate()` or `hunter_add_package()`?

I'll double check tomorrow though and see if something changed since last time I tried to compile with these blocks present.

ruslo replied:
> Is BUILD_SHARED_LIBS set to OFF by default inside HunterGate() or hunter_add_package()?

Yes, but it's not really related. `option` is customizable feature, user can always control it from outside. If you use `option` in code you simply declaring default value hence there is no need to remove it:
* https://cgold.readthedocs.io/en/latest/tutorials/variables/cache.html#use-case

ruslo replied:
☝️

lsolanka-seebyte replied:
Thanks for the link, I actually didn't know the `option` works as a cached variable.

The issue I have with what you want me to do is that the initial value for `BUILD_SHARED_LIBS` above is set to `ON`, which forces the shared libs. I am not aware where Hunter really sets it to OFF by default because when I try to recompile with the option above (when reverting the patch) all libraries are built as .so - in rospp for instance.

Can you perhaps just confirm that you want to have this option's initial value set to `ON` as in the code above? I'm happy to keep the option but I think it should be set to `OFF`, i.e. no initial value. Otherwise that will force us to include a BUILD_SHARED_LIBS=OFF line to in hunter.cmake in each package that will come?

lsolanka-seebyte replied:
> Is BUILD_SHARED_LIBS set to OFF by default inside HunterGate() or hunter_add_package()?
> 
> Yes, but it's not really related. option is customizable feature, user can always control it from outside. If you use option in code you simply declaring default value hence there is no need to remove it:
> 
>     https://cgold.readthedocs.io/en/latest/tutorials/variables/cache.html#use-case
Ok, taking back what I said above. I see how Hunter creates a cache file in which `BUILD_SHARED_LIBS=OFF`. So all dependencies of a parent project will be built as static, unless overridden.

However, if the user decides to have the parent project to remain a catkin_package by having the usual ROS commands (simplified example for `roscpp`):

```
project(roscpp)

hunter_add_package(ros-catkin)
hunter_add_package(ros-roscpp_core)
# more dependencies

find_package(catkin REQUIRED COMPONENTS
  cpp_common message_generation rosconsole roscpp_serialization roscpp_traits
  rosgraph_msgs rostime std_msgs xmlrpcpp
)

catkin_package(
  # INCLUDE_DIRS ...
  # LIBRARIES roscpp ${PTHREAD_LIB}
  # CATKIN_DEPENDS cpp_common message_runtime ...
  # DEPENDS "Boost COMPONENTS chrono filesystem signals system"
)
```

Then this will include `windows.cmake` and `library.cmake` into the parent project, which in turn will build the parent project with shared libs. This would be a bit unexpected for someone who is building with Hunter and expecting the project to have static libs perhaps?

Let me know what you think, I'm happy to go with whatever you think is best but the above happens when I actually build a ROS application and the options are set to initial value ON.

Sorry for a long comment for such a small thing...

ruslo replied:
> Then this will include windows.cmake and library.cmake into the parent project, which in turn will build the parent project with shared libs

Will not be a problem if user will do `option(BUILD_SHARED_LIBS "...")` before but I agree that it's pretty surprising behavior, I'm okay with removing it.

ruslo commented on cmake/tools/libraries.cmake:
```diff
@@ -7,11 +7,6 @@
 
-# Make sure this is already defined as a cached variable (@sa platform/windows.cmake)
-if(NOT DEFINED BUILD_SHARED_LIBS)
-  option(BUILD_SHARED_LIBS "Build dynamically-linked binaries" ON)
```
`option` should work fine, no need to remove it

ruslo replied:
☝️

lsolanka-seebyte replied:
See my reply for the other file.

ruslo commented on cmake/test/gtest.cmake:
```diff
@@ -284,121 +284,17 @@ function(catkin_find_google_test_source include_path src_path
   endif()
 endfunction()
 
-find_package(GMock QUIET)
```
The best approach would be to create minimal patch to simplify updates in future. I think instead of removing the whole block you can just do:
```cmake
if(HUNTER_ENABLED)
  hunter_add_package(...)
  # ....
endif()

if(NOT HUNTER_ENABLED)
  find_package(GMock QUIET)
endif()

if(NOT GMOCK_FOUND AND NOT HUNTER_ENABLED)
  # ...
elseif(NOT HUNTER_ENABLED)
  # ...
endif()
```

Also assuming that this code used for testing and we don't need tests for Hunter may be it make sense to not to touch this block at all and disable testing by some option.

lsolanka-seebyte replied:
Thanks for this, it's a good idea to add the if() blocks. I have simplified the diff as you're suggesting.

> Also assuming that this code used for testing and we don't need tests for Hunter may be it make sense to not to touch this block at all and disable testing by some option.

The functions catkin_add_gtest() and catkin_add_gmock() are used both for testing the core ROS packages, but also by any other packages, and the usage of these functions is actually quite prevalent. I think it's worth pulling GTest and GMock through Hunter, so that these are available for the downstream Hunterised projects for running the tests without them having to make any patches to the already existing code that use catkin_add_gtest/gmock().

lsolanka-seebyte replied:
There is a CATKIN_ENABLE_TESTING option which if set to OFF will disable building any tests. This can be put to hunter.cmake for each package.

ruslo replied:
> This can be put to hunter.cmake for each package

👍

lsolanka committed:
minimize gtest.cmake diff even more

Instead of removing large sections of code, to minimize potential
conflicts, if() statements are added.

lsolanka-seebyte commented on cmake/test/gtest.cmake:
```diff
@@ -397,7 +416,7 @@ else()
 
 # For Visual C++, need to increase variadic template size to build gtest
-if(GTEST_FOUND)
+if(GTEST_FOUND AND NOT HUNTER_ENABLED)
```
I have re-instated this block but only for HUNTER_ENABLED=OFF. I assume this is not necessary when using Hunter - all seems to compile fine without it, but please let me know if this should be unconditional.

ruslo replied:
It seem this change is not needed.

ruslo replied:
☝️

ruslo commented on cmake/test/gtest.cmake:
```diff
@@ -284,8 +284,27 @@ function(catkin_find_google_test_source include_path src_path
+if(HUNTER_ENABLED)
+    hunter_add_package(GTest)
+
+    if(NOT EXISTS "${GTest_LICENSES}")
```
This code can be removed if you're not really using `GTest_LICENSES` variable.

ruslo replied:
☝️

ruslo commented on cmake/templates/pkgConfig.cmake.in:
```diff
@@ -116,7 +116,39 @@ if(NOT "@PROJECT_CMAKE_CONFIG_INCLUDE_DIRS@ " STREQUAL " ")
 endif()
 
-set(libraries "@PKG_CONFIG_LIBRARIES@")
+set(depends "@PROJECT_DEPENDENCIES@")
```
Minimizing patch goal: this code can be moved lower.

ruslo replied:
☝️

lsolanka-seebyte replied:
Ok, I've moved this down and therefore minimized removed lines. However, as a side effect I had to add some extra code to traverse the libraries that are exported by the project (and which will be added to ${catkin_LIBRARIES} of downstream projects). The diff is slightly smaller but not that much.

I'm afraid that's as much as I can do, in order to not break code where a package has a dependency a library that needs to be found using find_library().

ruslo commented on cmake/catkin_package.cmake:
```diff
@@ -134,46 +134,7 @@ function(_catkin_package)
-
-  # for catkin packages it can be guaranteed that they are find_package()-able and have pkg-config files
-  set(PROJECT_DEPENDENCIES "")
+  set(PROJECT_DEPENDENCIES ${PROJECT_DEPENDS})
```
Minimizing patch goal: can we just add some code to nullify the effect of `foreach(depend ${PROJECT_DEPENDS})` loop instead of removing it completely?

lsolanka-seebyte replied:
Done. Thanks for that. I had to add some nullify statements, otherwise absolute paths to libraries would be propagated inside `LINK_INTERFACE_LIBRARIES` of exported targets (see the new diff).

lsolanka committed:
simplify catkin_package.cmake patch

It is not necessary to remove the block in the patch and therefore it
has been re-instated. However, the effect of setting
PROJECT_DEPENDENCIES_LIBRARIES and _INCLUDE_DIRS has to be nullified,
otherwise these will be propagated to link libraries of exported
targets.

lsolanka committed:
revert condition on variadic template size

lsolanka committed:
remove GTest_LICENSES blocks

lsolanka committed:
try to minimize dependency resolution diff

lsolanka-seebyte commented:
I think I have addressed all comments except the `BUILD_SHARED_LIBS` where I just want to get some clarification on the initial value.

For patch minimization: I have provided explanation above for each change that I think is necessary in order to keep consistency with ROS conventions, and especially the usage of find_library in various ROS packages that use catkin.

Let me know if you see some further ways how to minimize the patch or if you have any further comments.

lsolanka-seebyte commented:
Also let me know if I missed something.

ruslo commented:
Released: https://github.com/hunter-packages/catkin/releases/tag/v0.7.11-p0
